### ISSUE TITLE
[BUG] Soundness issue when a `'a Living_core.t` gets garbage collected

### ISSUE THREAD

**Issue by Fizzixnerd** [Fizzixnerd/ocaml-living#3]
Consider the following:

```ocaml
  type t
  let s : t Ctypes.structure Ctypes.typ = Ctypes.structure "dummy"
  let x_f = Ctypes.field s "x" (Ctypes.ptr Ctypes.int)
  let () = Ctypes.seal s

  (** [r] is uninitialized here *)
  let r = allocate_n ~count:1 s

  (** [f : unit -> unit Living_core.t] writes to [r]. *)
  let f () =
    let open Living_core.Let_syntax in
    let open Living_ctypes in
    let* x' = r |-> x_f in
    let* () = x' <-@ allocate int 7 in
    Living_core.return ()

  (** [g: unit -> int Living_core.t] reads from [r]. *)
  let g () =
    let open Living_core.Let_syntax in
    let open Living_ctypes in
    let* x' = r |-> x_f in
    Living_core.bind (!@) (!@ x')

  (** Main program; composes f and g in various ways and calls the GC after each run. *)
  let () =
    let open Living_core in
    let _ = 
      (* No problems, [r] is global and is uninitialized *)
      Gc.compact ();
      let* x = f () >>= g in
      Printf.printf "seven: %d\n" x;
      Living_core.return ()
    in

    let _ =
      (* Problem! this will GC the allocated pointer *)
      Gc.compact ();
      let* x = g () in
      Printf.printf "Garbage: %d" x
      Living_core.return ()
    in
    ()
```

**Comment by Fizzixnerd** [comment]
The issue seems to arise when we allow a `Living_core.t` _itself_ get garbage collected without meaning to.

## Solutions

If we had linear types, we could ensure linear only access to `r` and ensure the consumption of all `Living_core.t`s which would obviate the problem.

We do not have linear types, so we have to go with the next best thing.

The next best thing _I_ can think of is the following: copy the `Ctypes` handling of dynamic function pointers, by attaching a finalizer to every `Living_core.t` and warning that it's fallen out of scope, and then preventing it from being collected.  In this way, we ensure soundness is not violated and you won't get use-after-free bugs.

To compensate for when you really do want to drop a `Living_core.t`, we provide a `Living_core.free` function that removes this finalizer, and returns the `unsafe_value`.  We will then make the type abstract to prevent misuse of `unsafe_value` without freeing.

The idea is you stay in the monad if you want soundness, and only start dropping things when you're sure you need to optimize for memory consumption, at the risk of making a mistake and losing soundness.

**Comment by Fizzixnerd** [comment]
Fixed in 92d69fd
